### PR TITLE
[SMTChecker] Add missing constraints for block properties

### LINKED ISSUE #10659: [SMTChecker] Unknown exception during compilation: map::at
## Description

This contract:
```solidity
pragma experimental SMTChecker;
contract C {
    function f() public returns (uint) {
        return block.chainid;
    }
}
```

produces:

```
Unknown exception during compilation: map::at
```

when compiled with solc.

On master, using AFL fuzzing.  Another discovery using https://github.com/agroce/afl-compiler-fuzzer.

## Environment

- Compiler version:  0.8.1-develop.2020.12.18+commit.72efdd3e.mod.Linux.clang
- Target EVM version (as per compiler settings): N/A
- Framework/IDE (e.g. Truffle or Remix): N/A
- EVM execution environment / backend / blockchain client: N/A
- Operating system: Ubuntu 20.04 in docker

## Steps to Reproduce

Above shows pretty clearly, I think.

### LINKED ISSUE #10693: missing constraints on block.timestamp in SMTChecker
In the following code:
```
pragma experimental SMTChecker;

contract C { 
    function f() public returns (uint) {
        uint b = block.timestamp;
        uint a = b + 0;
        return a;
    }
}
```
Compiling with `0.7.6`, both bmc and chc reports that `b + 0` results in an overflow when `b = 2**256`. However, this is not a possible value for `block.timestamp`. Here's the error message from bmc:
```
Warning: BMC: Overflow (resulting value larger than 2**256 - 1) happens here.
 --> ~/c.sol:6:18:
  |
6 |         uint a = b + 0;
  |                  ^^^^^
Note: Counterexample:
   = 0
  <result> = 2**256
  a = 0
  b = 2**256
```

I'm also confused by the loose hanging `= 0`, I think maybe `block.timestamp` isn't pretty printed properly...

### PR DESCRIPTION
**Pull request by blishko**
This PR adds missing support for `block.chainid` and adds missing constraints on block properties `chainid`, `difficulty`, `gaslimit`, `number` and `timestamp`.

Fixes #10693.
Fixes #10659.

### CODE REVIEW

File: libsolidity/formal/BMC.cpp [hunk 1]
```diff
@@ -158,6 +158,7 @@ bool BMC::visit(FunctionDefinition const& _function)
 	{
 		reset();
 		initFunction(_function);
+		m_context.addAssertion(m_context.state().txConstraints(_function));
```
Comment (blishko) [inline libsolidity/formal/BMC.cpp:161]: Not sure if we should override `initFunction` for BMC, or this is sufficient.

Comment (leonardoalt) [reply libsolidity/formal/BMC.cpp:161]: I think this is fine

**Review by leonardoalt — COMMENTED** [review]
Looks good, but please add a changelog entry.

**Comment by leonardoalt** [comment]
Looks like a few tests will need updates, + the cli tests

**Comment by blishko** [comment]
Yes, I will add a Changelog and update the tests. Unfortunately, this is another change in the encoding that causes Z3 to behave differently. Also, there is quite a number of differences between the Linux in CLI and my local OSX.

**Comment by leonardoalt** [comment]
CLI result changes?

**Comment by blishko** [comment]
> CLI result changes?

No, sorry, I meant CI... The differences are only in SMTChecker tests.

File: Changelog.md [hunk 2]
```diff
@@ -11,6 +11,8 @@ Bugfixes:
  * Code Generator: Fix length check when decoding malformed error data in catch clause.
  * SMTChecker: Fix false negatives in overriding modifiers and functions.
  * SMTChecker: Fix false negatives when analyzing external function calls.
+ * SMTChecker: Fix missing constraints on block properties.
```
Comment (leonardoalt) [inline Changelog.md]:
```suggestion
 * SMTChecker: Fix missing type constraints on block variables.
```

File: Changelog.md [hunk 3]
```diff
@@ -11,6 +11,8 @@ Bugfixes:
...
+ * SMTChecker: Fix internal error on `block.chainid`.
```
Comment (leonardoalt) [inline Changelog.md]:
```suggestion
 * SMTChecker: Fix internal error on ``block.chainid``.
```

File: Changelog.md [hunk 4]
```diff
@@ -11,6 +11,8 @@ Bugfixes:
  * Code Generator: Fix length check when decoding malformed error data in catch clause.
  * SMTChecker: Fix false negatives in overriding modifiers and functions.
  * SMTChecker: Fix false negatives when analyzing external function calls.
+ * SMTChecker: Fix missing type constraints on block variable.
```
Comment (leonardoalt) [inline Changelog.md]:
```suggestion
 * SMTChecker: Fix missing type constraints for block variables.
```
I mean the `block.*` variables

Comment (blishko) [reply Changelog.md]: Should I formulate it as `block.*` or just "block variables", as in the suggestion?

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**